File: Module1 - Zachary Abbe -CS499/Project1-ZacharyAbbe-CS320/TaskService.java
/*****************************
 * Name: Zachary Abbe
 * Class: CS320
 * Date: 04/12/2025
 * Desc: This contains the TaskService and maintains the tasks list and can add, update and delete tasks
 ****************************/
import java.util.ArrayList;

public class TaskService {
    public ArrayList<Task> tasklist = new ArrayList<Task>();

    //creates a task object and adds it to the Array
    public boolean addTask(Task task) {
        String id = task.getUniqId();

        // Check if task with the same ID already exists
        for (Task t : tasklist) {
            if (t.getUniqId().equalsIgnoreCase(id)) {
                System.out.println("Task with ID " + id + " already exists.");
                return false;
            }
        }

        // If ID is unique, add task to the list
        tasklist.add(task);
        return true;
    }




    //task finder by id
    public Task findTaskbyId(String id) {
        for (Task t : tasklist) {
            if (t.getUniqId().equalsIgnoreCase(id)) {
                return t;
            }
        }
        return null;
    }
    //delete task from the tasklist
    public void deleteTask(String id){
        tasklist.removeIf(t->t.getUniqId().equalsIgnoreCase(id));
    }

    //updates the Task name if id is found
     public void updateName(String id, String name){
        Task task = findTaskbyId(id);
        if (task != null){
            task.setFullName(name);
        }
        else{
            throw new IllegalArgumentException("Task not found - can't update name");
        }
     }

     //updates the task description if the id is foind
     public void updateDesc(String id, String desc){
        Task task = findTaskbyId(id);
        if (task != null){
            task.setDesc(desc);
        }
        else{
            throw new IllegalArgumentException("Task not found - can't update the description");
        }
     }

    // Getter method for test verification
    public ArrayList<Task> getTaskList() {
        return tasklist;
    }
}

File: CS499Project1and2-ZacharyAbbe/TaskService.cpp
#include "TaskService.h"
#include <algorithm>
#include <stdexcept>

using namespace std;

bool TaskService::addTask(const Task& task) 
{
    /*for (const auto& t : tasklist) {
        if (t.getUniqId() == task.getUniqId()) 
        {
            cout << "Task with ID " << task.getUniqId() << " already exists." << endl;
            return false;
        }
    }
    tasklist.push_back(task);
    return true;*/
    if (taskList.count(task.getUniqId())) {
        cout << "Task with ID " << task.getUniqId() << " already exists." << endl;
        return false;
    }
    taskList.insert({ task.getUniqId(), task });
    return true;
}

Task* TaskService::findTaskbyId(const string& id)
{
    //for (auto& t : tasklist) 
    //{
      //  if (t.getUniqId() == id) 
       // {
         //   return &t;
        //}
    //}
    // we will use the new method for finding the ID using our O(1) method
    auto it = taskList.find(id);
    if (it != taskList.end())
    {
        return &it->second;
    }
    return nullptr;
}

void TaskService::deleteTask(const string& id) 
{// again we have no need for the complicated lambda function to delete
    //tasklist.erase(remove_if(tasklist.begin(), tasklist.end(),[&](const Task& t) { return t.getUniqId() == id;}), tasklist.end());
    taskList.erase(id);
}

void TaskService::updateName(const string& id, const string& name) 
{
    Task* task = findTaskbyId(id);
    if (task) 
    {
        task->setFullName(name);
    }
    else 
    {
        throw invalid_argument("Task not found - can't update name");
    }
}

void TaskService::updateDesc(const string& id, const string& desc) 
{
    Task* task = findTaskbyId(id);
    if (task) 
    {
        task->setDesc(desc);
    }
    else 
    {
        throw invalid_argument("Task not found - can't update the description");
    }
}

vector<Task> TaskService::getTaskList() const 
{
    vector<Task> tasks;
    for (const auto& pair : taskList)
    {
        tasks.push_back(pair.second);
    }
    return tasks;
}